### PR TITLE
Add SFT training features and fix SFT pipeline issues

### PR DESCRIPTION
**Pull request by tanzelin430**
## Summary

This PR adds useful features for long-running training jobs and fixes some issues in the SFT pipeline.

---

### 1. Add automatic checkpoint cleanup (`max_ckpt_to_keep`)

Long training runs can generate many checkpoints and exhaust disk space. This adds a built-in mechanism to limit checkpoint count.

**Changes:**
- `roll/configs/base_config.py`: New `max_ckpt_to_keep` parameter
- `roll/pipeline/base_pipeline.py`: New `_cleanup_old_checkpoints()` method

**Usage:**
```yaml
save_steps: 100
max_ckpt_to_keep: 3  # Keep only the latest 3 checkpoints, 0 = keep all
```

**How It Works:**
1. After each checkpoint save, scan `output_dir` for all subdirectories
2. Collect all `checkpoint-{step}` directories
3. Sort by step number, keep the latest N, delete older ones

---

### 2. Add wandb offline mode support

Training on clusters without internet access cannot use wandb logging. Added `mode` parameter to `WandbTracker` to support offline logging.
image

**Usage:**
```yaml
tracker_kwargs:
  mode: offline  # Can be "online", "offline", or "disabled"
```

---

### 3. Add tqdm progress bar for SFT training

Added tqdm progress bar to `sft_pipeline.py` for better training visualization.

---

### 4. Add `setup.py` for pip installation

Added `setup.py` to support editable installation:

```bash
cd ROLL
pip install -e .
```

---

### 5. Fix loss function call in `sft_worker.py`

The original code calls `op_compute_language_loss(output_tensor, labels)`, but `output_tensor` is actually logits, while `op_compute_language_loss` expects pre-computed per-token losses.

**Fix:** Use `op_compute_language_loss_from_logits` which correctly computes cross-entropy loss from logits.

```python
def loss_func(self, data: DataProto, output_tensor: torch.Tensor):
    labels = data.batch["labels"]
    loss, _ = self.strategy.op_compute_language_loss_from_logits(output_tensor, labels)
    metrics = {f"{self.worker_config.name}/loss": loss.detach().float().unsqueeze(0)}
    return loss, metrics
```

Note: `.float()` conversion is needed because bf16 tensors don't support numpy conversion.

---

### 6. Fix double-shift issue in `sft_pipeline.py`

After applying fix #5, labels get shifted twice:
1. `DataCollatorForSFT` with `shift_feature=True` (default)
2. Inside `op_compute_language_loss_from_logits`

This causes misalignment between logits and labels.

**Fix:** Set `shift_feature=False` since the loss function already handles shifting:

```python
data_collator = DataCollatorForSFT(
    ...
    shift_feature=False,  # loss func already does shift
)
```

---

## Files Changed

| File | Changes |
|------|---------|
| `roll/configs/base_config.py` | Add `max_ckpt_to_keep` parameter |
| `roll/pipeline/base_pipeline.py` | Add checkpoint cleanup |
| `roll/utils/tracking.py` | Add wandb offline mode support |
| `roll/pipeline/sft/sft_worker.py` | Fix loss computation |
| `roll/pipeline/sft/sft_pipeline.py` | Fix double-shift issue, add tqdm progress bar |
| `setup.py` | New file for pip install |

---

## Breaking Changes

None. All changes are backward compatible.

### CODE REVIEW

**Comment by xuehuanran** [comment]
Hi, thanks for your interest in the ROLL.
- Features 1-4 are excellent suggestions. We will look into them.  
- For bug 5 and 6, In the SFT pipeline, the output_tensor is actually the loss. This happens because the SFT pipeline provides labels for the computation in the `MegatronStrage`, so Megatron returns the language loss. Other pipelines set labels to `None`, which returns logits.
Here is the related code:
https://github.com/alibaba/ROLL/blob/main/roll/distributed/strategy/megatron_strategy.py#L428
```python
def inner_forward_step(self, loss_func, data_iterator: Iterator[DataProto], model):
        data = next(data_iterator)
        input_ids = data.batch["input_ids"]
        attention_mask = data.batch["attention_mask"]
        labels = data.batch["labels"] if "labels" in data.batch else None  # labels is only used for sft
        ...
        # output_tensor is `logits` or `loss`
        output_tensor = model(
            input_ids=input_ids, attention_mask=attention_mask, position_ids=position_ids, labels=labels,
            packed_seq_params=packed_seq_params, **forward_args
        )
```

**Comment by tanzelin430** [comment]
@xuehuanran Thanks for quick reply. Actually I did not use megatron but deepseed as my backend (Because of some env problems that are hard to solve on my dev machine). So I think maybe we should use different function for loss compute based on different backend?

**Comment by xuehuanran** [comment]
@tanzelin430 
Sorry, the SFT pipeline is currently not tested on the DeepSpeed strategy.

> So I think maybe we should use different function for loss compute based on different backend?

Your understanding is right, it's typically not needed to modify the `loss_func` implementation within the `Worker` for different backends, as the loss computation in the `Worker` is general. Instead, the loss computation method specific to each backend should be implemented in the corresponding `Strategy` class.

Therefore, you can move the loss implementation to the `DeepSpeedStrategy`. We will also be working on integrating SFT with other strategies like DeepSpeed and FSDP in the future.

**Comment by tanzelin430** [comment]
@xuehuanran Maybe I can help to add a quick fix for this problem? Do you support outside PR?
Besides, I am going to use ROLL for RL in my science project, also with deepseed backend. Have you test the RL pipeline on deepspeed?

**Comment by xuehuanran** [comment]
@tanzelin430 Sure, look forward to the PR; The deepspeed strategy is tested on the RLVR and Agentic pipeline.
